File: finanze/domain/transactions.py
from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic.dataclasses import dataclass

from domain.base import BaseData
from domain.dezimal import Dezimal
from domain.entity import Entity
from domain.global_position import ProductType


class TxType(str, Enum):
    BUY = "BUY"
    SELL = "SELL"
    DIVIDEND = "DIVIDEND"
    RIGHT_ISSUE = "RIGHT_ISSUE"
    RIGHT_SELL = "RIGHT_SELL"
    SUBSCRIPTION = "SUBSCRIPTION"
    SWAP_FROM = "SWAP_FROM"
    SWAP_TO = "SWAP_TO"

    INVESTMENT = "INVESTMENT"
    REPAYMENT = "REPAYMENT"
    INTEREST = "INTEREST"


@dataclass
class BaseTx(BaseData):
    id: UUID
    ref: str
    name: str
    amount: Dezimal
    currency: str
    type: TxType
    date: datetime
    entity: Entity
    is_real: bool
    product_type: ProductType


@dataclass
class BaseInvestmentTx(BaseTx):
    pass


@dataclass
class AccountTx(BaseTx):
    fees: Dezimal
    retentions: Dezimal
    interest_rate: Optional[Dezimal] = None
    avg_balance: Optional[Dezimal] = None


@dataclass
class StockTx(BaseInvestmentTx):
    net_amount: Dezimal
    isin: Optional[str]
    shares: Dezimal
    price: Dezimal
    fees: Dezimal
    ticker: Optional[str] = None
    market: Optional[str] = None
    retentions: Optional[Dezimal] = None
    order_date: Optional[datetime] = None
    linked_tx: Optional[str] = None


@dataclass
class FundTx(BaseInvestmentTx):
    net_amount: Dezimal
    isin: str
    shares: Dezimal
    price: Dezimal
    market: str
    fees: Dezimal
    retentions: Optional[Dezimal] = None
    order_date: Optional[datetime] = None


@dataclass
class FactoringTx(BaseInvestmentTx):
    net_amount: Dezimal
    fees: Dezimal
    retentions: Dezimal
    interests: Dezimal


@dataclass
class RealStateCFTx(BaseInvestmentTx):
    net_amount: Dezimal
    fees: Dezimal
    retentions: Dezimal
    interests: Dezimal


@dataclass
class DepositTx(BaseInvestmentTx):
    net_amount: Dezimal
    fees: Dezimal
    retentions: Dezimal
    interests: Dezimal


@dataclass
class Transactions:
    investment: Optional[list[BaseInvestmentTx]] = None
    account: Optional[list[AccountTx]] = None

    def __add__(self, other):
        investment = (self.investment or []) + (other.investment or [])
        account = (self.account or []) + (other.account or [])
        return Transactions(investment=investment, account=account)


@dataclass
class TransactionsResult:
    transactions: list[BaseTx]


@dataclass
class TransactionQueryRequest:
    page: int = 1
    limit: int = 10
    entities: Optional[list[UUID]] = None
    excluded_entities: Optional[list[UUID]] = None
    product_types: Optional[list[ProductType]] = None
    from_date: Optional[datetime] = None
    to_date: Optional[datetime] = None
    types: Optional[list[TxType]] = None

File: mobile/src/domain/core/transactions.ts
import { Entity } from "./entity"
import { DataSource } from "./fetchRecord"
import { EquityType, FundType, ProductType } from "./globalPosition"
import { Dezimal } from "@/domain/dezimal"

export enum TxType {
  BUY = "BUY",
  SELL = "SELL",
  DIVIDEND = "DIVIDEND",
  RIGHT_ISSUE = "RIGHT_ISSUE",
  RIGHT_SELL = "RIGHT_SELL",
  SUBSCRIPTION = "SUBSCRIPTION",
  SWAP_FROM = "SWAP_FROM",
  SWAP_TO = "SWAP_TO",
  TRANSFER_IN = "TRANSFER_IN",
  TRANSFER_OUT = "TRANSFER_OUT",
  SWITCH_FROM = "SWITCH_FROM",
  SWITCH_TO = "SWITCH_TO",
  INVESTMENT = "INVESTMENT",
  REPAYMENT = "REPAYMENT",
  INTEREST = "INTEREST",
  FEE = "FEE",
}

export interface BaseTx {
  id: string | null
  ref: string
  name: string
  amount: Dezimal
  currency: string
  type: TxType
  date: string
  entity: Entity
  source: DataSource
  productType: ProductType
}

export interface BaseInvestmentTx extends BaseTx {}

export interface AccountTx extends BaseTx {
  fees: Dezimal
  retentions: Dezimal
  interestRate?: Dezimal | null
  avgBalance?: Dezimal | null
  netAmount?: Dezimal | null
}

export interface StockTx extends BaseInvestmentTx {
  shares: Dezimal
  price: Dezimal
  fees: Dezimal
  netAmount?: Dezimal | null
  isin?: string | null
  ticker?: string | null
  market?: string | null
  retentions?: Dezimal | null
  orderDate?: string | null
  linkedTx?: string | null
  equityType?: EquityType | null
}

export interface CryptoCurrencyTx extends BaseInvestmentTx {
  currencyAmount: Dezimal
  symbol: string
  price: Dezimal
  fees: Dezimal
  contractAddress?: string | null
  netAmount?: Dezimal | null
  retentions?: Dezimal | null
  orderDate?: string | null
}

export interface FundTx extends BaseInvestmentTx {
  shares: Dezimal
  price: Dezimal
  fees: Dezimal
  netAmount?: Dezimal | null
  isin?: string | null
  market?: string | null
  retentions?: Dezimal | null
  orderDate?: string | null
  fundType?: FundType | null
}

export interface FundPortfolioTx extends BaseInvestmentTx {
  portfolioName: string
  iban?: string | null
  fees?: Dezimal
}

export interface FactoringTx extends BaseInvestmentTx {
  fees: Dezimal
  retentions: Dezimal
  netAmount?: Dezimal | null
}

export interface RealEstateCFTx extends BaseInvestmentTx {
  fees: Dezimal
  retentions: Dezimal
  netAmount?: Dezimal | null
}

export interface DepositTx extends BaseInvestmentTx {
  fees: Dezimal
  retentions: Dezimal
  netAmount?: Dezimal | null
}

export interface Transactions {
  investment?: BaseInvestmentTx[] | null
  account?: AccountTx[] | null
}

export interface TransactionsResult {
  transactions: BaseTx[]
}

export interface TransactionQueryRequest {
  page?: number
  limit?: number
  entities?: string[] | null
  excludedEntities?: string[] | null
  productTypes?: ProductType[] | null
  fromDate?: string | null
  toDate?: string | null
  types?: TxType[] | null
  historicEntryId?: string | null
}
